core/src/Cache/CacheFallbackOperations.cs

```diff
@@ -186,16 +206,25 @@ internal class CacheFallbackOperations
         /// (nor will GetAccounts / RemoveAccount work)
         /// 
         /// </summary>
-        public static void RemoveAdalUser(ILegacyCachePersistance legacyCachePersistance,
-            string displayableId, ISet<string> environmentAliases, string identifier)
+        public static void RemoveAdalUser(
+            ILegacyCachePersistance legacyCachePersistance,
+            ISet<string> environmentAliases,
+            string displayableId,
+            string accountOrUserId)
         {
             try
             {
                 IDictionary<AdalTokenCacheKey, AdalResultWrapper> adalCache =
                 AdalCacheOperations.Deserialize(legacyCachePersistance.LoadCache());
 
-                RemoveEntriesWithMatchingId(environmentAliases, identifier, adalCache);
-                RemoveEntriesWithMatchingName(environmentAliases, displayableId, adalCache);
+                if (!string.IsNullOrEmpty(accountOrUserId))
```

Comment:
We are now deleting too much. The logic for deleting is documented in the md file and in the comments of this method.  
@jennyf19 - I was under the impression that the original code had this IF statement in here as well. Did you find use cases where we have to remove based on account_id AND based on username as well? I was pretty convinced that if we have an account_id we can stop there.

Reply:
Petro asked us to do removal both based on account and on displayable id.

---
In reply to: [219186124](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1259#discussion_r219186124) [](ancestors = 219186124)

Reply:
I don't think the current implementation is ok. It deletes too much. I could remove line 197 (the call to RemoveEntriesWithMatchingId) and get the same result because  RemoveEntriesWithMatchingName ignores the client id althogheter. 

For example in the scenario where: 

a. Several users for which we do not have preferredUsername get tokens
b. One user activates a token cache removal (e.g. signs out)

Actual: all users get signed out of the adal cache 


I think maybe what Petro was trying to say is to scope the deletion both by accountId and by username at the same time ?

Reply:
if all tests passes we are fine...

---
In reply to: [219568047](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1259#discussion_r219568047) [](ancestors = 219568047)